src/stan/model/indexing/lvalue_varmat.hpp

```diff
@@ -0,0 +1,458 @@
+  stan::math::check_range("var_vector[uni] assign", name, x.size(),
+                          idxs.head_.n_);
+  const auto coeff_idx = idxs.head_.n_ - 1;
+  double prev_val = x.val().coeffRef(coeff_idx);
+  x.vi_->val_.coeffRef(coeff_idx) = y.val();
+  stan::math::reverse_pass_callback([x, y, coeff_idx, prev_val]() mutable {
+    x.vi_->val_.coeffRef(coeff_idx) = prev_val;
+    y.adj() += x.adj().coeffRef(coeff_idx);
+    x.adj().coeffRef(coeff_idx) = 0.0;
+  });
+}
+
+/**
+ * Assign to a non-contiguous subset of elements in a vector.
+ *
+ * Types:  vector[multi] <- vector
+ *
+ * @tparam Vec1 `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @tparam Vec2 `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @param[in] x Vector to be assigned.
+ * @param[in] idxs Index holding an `std::vector` of cells to assign to.
+ * @param[in] y Value vector.
+ * @param[in] name Name of variable (default "ANON").
+ * @param[in] depth Indexing depth (default 0).
+ * @throw std::out_of_range If any of the indices are out of bounds.
+ * @throw std::invalid_argument If the value size isn't the same as
+ * the indexed size.
+ */
+template <typename Vec1, typename Vec2,
+          require_all_var_vector_t<Vec1, Vec2>* = nullptr>
+inline void assign(Vec1&& x,
+                   const cons_index_list<index_multi, nil_index_list>& idxs,
+                   const Vec2& y, const char* name = "ANON", int depth = 0) {
+  stan::math::check_size_match("vector[multi] assign", "left hand side",
+                               idxs.head_.ns_.size(), name, y.size());
+  const auto x_size = x.size();
+  const auto assign_size = idxs.head_.ns_.size();
+  arena_t<std::vector<int>> x_idx(assign_size);
+  arena_t<Eigen::Matrix<double, -1, 1>> prev_vals(assign_size);
+  Eigen::Matrix<double, -1, 1> y_vals(assign_size);
+  std::unordered_set<int> x_set;
+  x_set.reserve(assign_size);
+  // We have to use two loops to avoid aliasing issues.
+  for (int i = assign_size - 1; i >= 0; --i) {
+    if (likely(x_set.insert(idxs.head_.ns_[i]).second)) {
+      stan::math::check_range("vector[multi] assign", name, x_size,
+                              idxs.head_.ns_[i]);
+      x_idx[i] = idxs.head_.ns_[i] - 1;
+      prev_vals.coeffRef(i) = x.vi_->val_.coeffRef(x_idx[i]);
+      y_vals.coeffRef(i) = y.vi_->val_.coeff(i);
+    } else {
+      x_idx[i] = -1;
+    }
+  }
+  for (int i = assign_size - 1; i >= 0; --i) {
+    if (likely(x_idx[i] != -1)) {
+      x.vi_->val_.coeffRef(x_idx[i]) = y_vals.coeff(i);
+    }
+  }
+
+  stan::math::reverse_pass_callback([x, y, x_idx, prev_vals]() mutable {
+    for (Eigen::Index i = 0; i < x_idx.size(); ++i) {
+      if (likely(x_idx[i] != -1)) {
+        x.vi_->val_.coeffRef(x_idx[i]) = prev_vals.coeffRef(i);
+        prev_vals.coeffRef(i) = x.adj().coeffRef(x_idx[i]);
+        x.adj().coeffRef(x_idx[i]) = 0.0;
```

Comment:
This line is writing adjoint, so it should be in the second loop.

Reply:
I think it needs to be in the first loop to handle values that overlap (unless I'm getting the pattern wrong). Because if we write to x and y in the same loop we are effectivly writing to X in both and I think the aliasing issue comes from that.

For like multi_vec

```
template <typename Vec>
void test_multi_alias_vec() {
  using stan::math::sum;
  using stan::math::var_value;
  using stan::model::test::check_vector_adjs;
  using stan::model::test::generate_linear_var_vector;
  auto x = generate_linear_var_vector<Eigen::VectorXd>(5, 1);
  Eigen::VectorXd x_val = x.val();
  vector<int> ns{1, 1, 2, 3};
  std::cout <<  "\nx val pre assign: \n" << x.val() << "\n";
  assign(x, index_list(index_multi(ns)), x.segment(1, 4));
  EXPECT_MATRIX_EQ(x.val().segment(0, 3), x_val.segment(2, 3));
  EXPECT_MATRIX_EQ(x.val().segment(3, 2), x_val.segment(3, 2));
  std::cout <<  "\nx val post assign: \n" << x.val() << "\n";
  stan::math::sum(x).grad();
  std::cout <<  "\nx adj: \n" << x.adj() << "\n";
  EXPECT_MATRIX_EQ(x.val(), x_val);
  Eigen::VectorXd exp_adj(5);
  exp_adj << 0, 0, 1, 2, 2;
  EXPECT_MATRIX_EQ(x.adj(), exp_adj);
}
```

We have 

```
x val pre assign: 
1
2
3
4
5

x val post assign: 
3
4
5
4
5
```

So if this was an `Eigen::Matrix<var>` it would be like cells {2, 3, 4} were put in cells {0, 1, 2}. So when we do `grad()` we should get back

```
x adj: 
0
0
1
2
2
```

Since the "var" in cells {0, 1} are overwritten their values would be zero, cell 2 overwites cell 0 so it's value should be one, and cells {3, 4} are used twice so their adjoints should be 2. But if we move 

```
x.adj().coeffRef(x_idx[i]) = 0.0;
```

To the second loop we get back

```
x adj: 
0
0
0
2
2
```

Reply:
If this is setting adjoints to zero in the first loop in can set some element to zero that is due to aliasing later read in the same loop. But you are right too. I believe we need to add a third loop just for setting adjoints to zero.

Reply:
oof yeah I think you are right.

Reply:
Actually idt we need a third loop.

`x_idx` in the below is going to only access unique values so it will look something like `{-1, 3, -1, 2, 1, 0}` or something of that ilk. 

```cpp
    for (Eigen::Index i = 0; i < x_idx.size(); ++i) {
      if (likely(x_idx[i] != -1)) {
        x.vi_->val_.coeffRef(x_idx[i]) = prev_vals.coeffRef(i);
        prev_vals.coeffRef(i) = x.adj().coeffRef(x_idx[i]);
        x.adj().coeffRef(x_idx[i]) = 0.0;
      }
    }
```

Reply:
You are right, I got confused here. This is fine.